include/envoy/network/io_handle.h

```diff
@@ -4,30 +4,108 @@
 
 #include "envoy/common/pure.h"
 
+#include "common/common/assert.h"
+
 namespace Envoy {
 namespace Network {
 
+// IoErrorCode::Again is used frequently. Define it to be a distinguishable address to avoid
+// frequent memory allocation of IoError instance.
+// If this is used, IoHandleCallResult has to be instantiated with a deleter that does not
+// deallocate memory for this error.
+#define ENVOY_ERROR_AGAIN reinterpret_cast<IoError*>(1)
+
+/**
+ * Base class for any I/O error.
+ */
+class IoError {
+public:
+  enum class IoErrorCode {
+    // No data available right now, try again later.
+    Again,
+    // Not supported.
+    NoSupport,
+    // Address family not supported.
+    AddressFamilyNoSupport,
+    // During non-blocking connect, the connection cannot be completed immediately.
+    InProgress,
+    // Permission denied.
+    Permission,
+    // Other error codes cannot be mapped to any one above in getErrorCode().
+    UnknownError
+  };
+  virtual ~IoError() {}
+
+  // Map platform specific error into IoErrorCode.
+  // Needed to hide errorCode() in case of ENVOY_ERROR_AGAIN.
+  static IoErrorCode getErrorCode(const IoError* err) {
+    ASSERT(err != nullptr);
```

Comment:
nit: Can we just make this function take a reference to avoid this assert? Same elsewhere in similar cases?

Reply:
I think `*ENVOY_ERROR_AGAIN` will cause crash.

Reply:
It won't crash actually, from the computer perspective a pointer and reference are the same thing, it won't crash until you actually try to access the memory (I'm 99% sure). I would go ahead and switch to a reference assuming it works.

Reply:
Yeah, you're right. I switched to use reference here.
If the caller try to dereference a nullptr, we will have seg fault. Is this more preferred than ASSERT?

Reply:
Yeah in general we prefer references for things that can't be null. In this case, I don't consider ENVOY_ERROR_AGAIN to be null.